internal/cache/timeoutCache_test.go

```diff
@@ -0,0 +1,112 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cache
+
+import (
+	"strconv"
+	"testing"
+	"time"
+)
+
+const (
+	testCacheTimeout = 100 * time.Millisecond
+)
+
+func (c *TimeoutCache) getForTesting(key interface{}) (*cacheEntry, bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	r, ok := c.cache[key]
+	return r, ok
+}
```

Comment:
Could we just use the exported API for testing.

Tests that I can think of:


TestCacheAdd()
  - adds an entry using Store/Add
  - removed the entry using Retrive/Remove and makes sure it gets the same one that was added

TestCacheTimeout
  - adds an entry
  - verifies the callback is invoked on expiry

TestCacheRemove()
  - adds an entry
  - removes it and verifies the callback is not invoked

TestCacheClear
  - add multiple entries
  - call Clear
  - make sure further none of the callbacks are invoked and that further calls to Remove return false

Reply:
I actually kind of like this, to define a test only method, and the method only exists during tests :)

This method is more of an implementation test, not behavior test. But I think it makes sense in unit tests like these.
Having an exported method gives other callers a change to check the internal state of the cache (we may need to add it later for testing purposes).

I also added a test TestCacheClear